File: MCTS_TicTacToe_cpp/Tree.cpp
//
// Created by João Vieira on 22/12/2019.
//

#include "Tree.h"
#include "Node.h"

Tree::Tree() {
    root = new Node();
}

Node *Tree::getRoot() {
    return root;
}

void Tree::setRoot(Node *root) {
    auto parent = root->getParent();

    if (parent)
        parent->getChildArray()->remove_if([root](Node *child)->bool{return child == root;});

    root->setParent(nullptr);
    delete this->root;
    this->root = root;
}

Tree::~Tree() {
    delete root;
}

File: MCTS_BlokusDuo_java/src/com/joaomiguelvieira/Tree.java
package com.joaomiguelvieira;

public class Tree {
    private Node root;

    public Tree() {
        root = new Node();
    }

    public Node getRoot() {
        return root;
    }

    public void setRoot(Node root) {
        this.root = root;

        Node tempNode = root.getParent();
        while (tempNode.getParent() != null) {
            tempNode.getChildArray().clear();
            tempNode = tempNode.getParent();
        }
    }
}
